kaggle_environments/envs/werewolf/werewolf.js

```diff
@@ -0,0 +1,4500 @@
+    (environment.info?.MODERATOR_OBSERVATION || []).forEach((stepEvents, kaggleStep) => {
+        (stepEvents || []).flat().forEach(dataEntry => {
+            const event = JSON.parse(dataEntry.json_str);
+            const dataType = dataEntry.data_type;
+            const visibleInUI = event.visible_in_ui ?? true;
+
+            console.log(`[RAW SEEN] Kaggle Step: ${kaggleStep}`, { dataType: dataType, event: event });
+
+            if (!visibleInUI) {
+              return;
+            }
+
+            // 1. Reset our memory on key phase-changing events.
+            if (event.event_name === 'day_start' || event.event_name === 'night_start' || event.description?.includes('Voting phase begins')) {
+                processedPhaseEvents.clear();
+            }
+
+            // 2. Generate a unique "fingerprint" based on essential event content.
+            let eventFingerprint = event.description;
+            const eventData = event.data;
+
+            // 3. Check against the new fingerprint.
+            if (processedPhaseEvents.has(eventFingerprint)) {
+                return; // This is a repeated state update, not a new event. Ignore it.
+            }
+            processedPhaseEvents.add(eventFingerprint);
+
+            // The rest of your original processing logic follows...
+            const isVisibleDataType = visibleEventDataTypes.has(dataType);
+            const isVisibleEntryType = systemEntryTypeSet.has(event.event_name);
+
+            if (!isVisibleDataType && !isVisibleEntryType) {
+                return;
+            }
+
+            // (Your existing code to push events to player.allEvents, player.displayEvents, etc.)
+            event.kaggleStep = kaggleStep;
+            event.dataType = dataType; // Ensure dataType is set
+            player.allEvents.push(event);
+            player.eventToKaggleStep.push(kaggleStep);
+
+            if (dataType !== 'PhaseDividerDataEntry') {
+                player.displayEvents.push(event);
+                player.displayStepToAllEventsIndex.push(allEventsIndex);
+                player.allEventsIndexToDisplayStep[allEventsIndex] = currentDisplayStep;
+                currentDisplayStep++;
+            }
+            allEventsIndex++;
+        });
+    });
+
+    console.log(`[FINAL STEP LIST]`, player.displayEvents);
+
+    const newSteps = player.displayEvents.map((event) => {
+        return player.originalSteps[event.kaggleStep];
+    });
+
+    setTimeout(() => {
+        if (window.kaggle) {
+            window.kaggle.environment.steps = newSteps;
+        }
+
+        console.log('context');
+        console.dir(context);
+
+        window.postMessage({ setSteps: newSteps }, "*");
+    }, 100); // A small delay to ensure player is ready
+    player.initialized = true;
+  }
+
+  // We patch the functions on the 'context' object directly.
+  if (context.__mainContext && !window.customPlayerControlsInjected) {
```

Comment:
not crazy about this mechanism for overwriting context - would be nice to figure out a more correct way to pass down whatever you need - happy to discuss.

Reply:
Yes, let's schedule sometime to go over possibilities. I am open to using a cleaner design.